### PR TITLE
fluentd: Added environment name as a record for all logs

### LINKED ISSUE #93: [2] Add cluster indentifier to logs
For multi-tenancy, we need to be able to associate each log entry with a specific workload cluster.
Part of #58

### PR DESCRIPTION
**Pull request by Xartos**
**What this PR does / why we need it**:

To support multi cluster we need to be able to differentiate between all logs in elasticsearch

**Which issue this PR fixes** *(use the format `fixes #<issue number>(, fixes #<issue_number>, ...)` to automatically close the issue when PR gets merged)*:

fixes #93 

**Public facing documentation PR** *(if applicable)*

**Special notes for reviewer**:

**Checklist:**

- [x] Added relevant notes to [WIP-CHANGELOG.md](https://github.com/elastisys/compliantkubernetes-apps/blob/main/WIP-CHANGELOG.md)
- [x] Proper commit message prefix on all commits
- [ ] Updated the [public facing documentation](https://github.com/elastisys/compliantkubernetes)
- Is this changeset backwards compatible for existing clusters? Applying:
    - [x] is completely transparent, will not impact the workload in any way.
    - [ ] requires running a migration script.
    - [ ] will create noticeable cluster degradation.
          E.g. logs or metrics are not being collected or Kubernetes API server
          will not be responding while upgrading.
    - [ ] requires draining and/or replacing nodes.
    - [ ] will change any APIs.
          E.g. removes or changes any CK8S config options or Kubernetes APIs.
    - [ ] will break the cluster.
          I.e. full cluster migration is required.

### CODE REVIEW

File: helmfile/values/fluentd-sc.yaml.gotmpl [hunk 1]
```diff
@@ -146,6 +146,13 @@ extraConfigMaps:
       </metric>
     </filter>
 
+    <filter **>
+      @type record_transformer
+      <record>
+        cluster.name "{{ .Values.global.environmentName }}"
```
Comment (Xartos) [inline helmfile/values/fluentd-sc.yaml.gotmpl:152]: Do we want to use `.Values.global.environmentName` or should we create a new entry like `.Values.global.clusterName`?

Comment (OlleLarsson) [reply helmfile/values/fluentd-sc.yaml.gotmpl:152]: I mean leaning towards a clusterName variable instead. 
Can't from the top of my head think of anywhere elase where we are using environment name than in the deploy scripts,  where it seems to be used to derive a cluster name :D 
https://github.com/elastisys/compliantkubernetes-apps/blob/d73f39d3afb022c804b027cfb9e30bb7b603f84e/scripts/deploy-wc.sh#L14

Comment (Xartos) [reply helmfile/values/fluentd-sc.yaml.gotmpl:152]: Seem to not be used anywhere else than there 
```bash
$ grep -r CLUSTER_NAME
scripts/deploy-wc.sh:export CLUSTER_NAME="${environment}_${cloud_provider}"
```

So I don't really see the benefit from using a separate variable (if we basically don't use this variable anywhere else). I can see that we want to rename `environmentName` to `clusterName` but I don't see the big benefit of having both

Comment (OlleLarsson) [reply helmfile/values/fluentd-sc.yaml.gotmpl:152]: Agree, rename `environmentName` to `clusterName`

**Review by lentzi90 — COMMENTED** [review]
This looks nice but I'm still a bit scared that I'm missing something.
To me the environment name was a common name for both/all clusters, whereas a cluster name should be unique for the cluster. So a `global.clusterName` doesn't really make sense to me as it isn't "global". But maybe I'm overthinking this?
To be clear, I think it works the way things are right now, but not when we get to multiple workload clusters or if we want logs also from SC.

**Comment by Xartos** [comment]
> This looks nice but I'm still a bit scared that I'm missing something.
> To me the environment name was a common name for both/all clusters, whereas a cluster name should be unique for the cluster. So a `global.clusterName` doesn't really make sense to me as it isn't "global". But maybe I'm overthinking this?
> To be clear, I think it works the way things are right now, but not when we get to multiple workload clusters or if we want logs also from SC.

Absolutely, maybe I could rephrase it to be more. 
"Removed `environmentName` variable (since we don't use it) and added new variable `clusterName`."

File: WIP-CHANGELOG.md [hunk 2]
```diff
@@ -38,6 +39,7 @@ The following options has been removed or replaced
 - Support for GCS
 - Backup retention for InfluxDB.
 - Add Okta as option for OIDC provider
+- Added record `cluster.name` in all logs to elasticsearch that matches the cluster setting `global.environmentName`
```
Comment (viktor-f) [inline WIP-CHANGELOG.md:42]:
```suggestion
- Added record `cluster.name` in all logs to elasticsearch that matches the cluster setting `global.clusterName`
```

**Review by viktor-f — CHANGES_REQUESTED** [review]
Changing to `clusterName` sounds good. But should we then also change `CK8S_ENVIRONMENT_NAME` to `CK8S_CLUSTER_NAME` to match? That would of course require changes in a few more places.

**Comment by Xartos** [comment]
> Changing to `clusterName` sounds good. But should we then also change `CK8S_ENVIRONMENT_NAME` to `CK8S_CLUSTER_NAME` to match? That would of course require changes in a few more places.

Good catch! I think that this is what we want to acheive from the multicluster in the end, but for now we init a whole "environment" not cluster by cluster. But maybe we could for now change `clusterName` to default to `${CK8S_ENVIRONMENT_NAME}-<wc|sc>` and when we change to init clusterwise we can rename this variable to `CK8S_CLUSER_NAME`

**Comment by viktor-f** [comment]
> > Changing to `clusterName` sounds good. But should we then also change `CK8S_ENVIRONMENT_NAME` to `CK8S_CLUSTER_NAME` to match? That would of course require changes in a few more places.
> 
> Good catch! I think that this is what we want to acheive from the multicluster in the end, but for now we init a whole "environment" not cluster by cluster. But maybe we could for now change `clusterName` to default to `${CK8S_ENVIRONMENT_NAME}-<wc|sc>` and when we change to init clusterwise we can rename this variable to `CK8S_CLUSER_NAME`

That sounds good :+1:

File: WIP-CHANGELOG.md (hunk 2 above)
Comment (Xartos) [reply WIP-CHANGELOG.md:56]: Since it's separate commits I thought this would go first and add the fluentd stuff and the next commit changes the name to clusterName

**Review by OlleLarsson — APPROVED** [review]

**Comment by OlleLarsson** [comment]
Should probably update our helper script https://github.com/elastisys/ck8s/blob/master/helper/configure-apps.bash#L119

**Comment by Xartos** [comment]
> Should probably update our helper script https://github.com/elastisys/ck8s/blob/master/helper/configure-apps.bash#L119

Already done ;) https://github.com/elastisys/ck8s/pull/665

**Comment by OlleLarsson** [comment]
> > Should probably update our helper script https://github.com/elastisys/ck8s/blob/master/helper/configure-apps.bash#L119
> 
> Already done ;) [elastisys/ck8s#665](https://github.com/elastisys/ck8s/pull/665)

Great :D

File: WIP-CHANGELOG.md (hunk 2 above)
Comment (lentzi90) [reply WIP-CHANGELOG.md:56]: While it makes sense when reviewing the commits, I think this makes the changelog confusing to read. I would change this line in the wip changelog also as part of the second commit. So you actually rename it everywhere, also in the changelog. :slightly_smiling_face:

Comment (Xartos) [reply WIP-CHANGELOG.md]: I changed the order of the two commits. I think it feels better to have them separate since it's two separate fixes. But I agree that with the order changed the WIP-CHANGELOG is more readable

**Review by viktor-f — COMMENTED** [review]
I think this looks good now, though I have not tested any of it.

One question regarding the log management: with this we would get all (kubernetes) logs in the same index even if we have multiple workload clusters, right? They would be identifiable since they have the `cluster.name` tag, so this is a fine solution. I'm just wondering if it would be even better to have them as separate indices. Just want to check if we have thought about that option.

**Comment by Xartos** [comment]
> I think this looks good now, though I have not tested any of it.
> 
> One question regarding the log management: with this we would get all (kubernetes) logs in the same index even if we have multiple workload clusters, right? They would be identifiable since they have the `cluster.name` tag, so this is a fine solution. I'm just wondering if it would be even better to have them as separate indices. Just want to check if we have thought about that option.

I don't know the pros/cons with that if we are talking "multi-workload-cluster" setup. But to me I think I would like to have the option to look at all logs in all cluster and to filter out if you want to. Maybe I'm not seeing the benefits of having it in its own index. But to me that is more if you want isolation between data.

**Comment by viktor-f** [comment]
> > I think this looks good now, though I have not tested any of it.
> > One question regarding the log management: with this we would get all (kubernetes) logs in the same index even if we have multiple workload clusters, right? They would be identifiable since they have the `cluster.name` tag, so this is a fine solution. I'm just wondering if it would be even better to have them as separate indices. Just want to check if we have thought about that option.
> 
> I don't know the pros/cons with that if we are talking "multi-workload-cluster" setup. But to me I think I would like to have the option to look at all logs in all cluster and to filter out if you want to. Maybe I'm not seeing the benefits of having it in its own index. But to me that is more if you want isolation between data.

Yes it's probably more convenient to have them in the same index. I think that we should probably stick with this for now, mostly wanted to note it as a possibility.
One advantage of separate indices is option to have different retention settings for different clusters. AFAIK you cannot do that with everything in the same index. In the same spirit, if one cluster started spamming logs it would not affect how much logs from other clusters are stored.

**Comment by Xartos** [comment]
> Yes it's probably more convenient to have them in the same index. I think that we should probably stick with this for now, mostly wanted to note it as a possibility.
> One advantage of separate indices is option to have different retention settings for different clusters. AFAIK you cannot do that with everything in the same index. In the same spirit, if one cluster started spamming logs it would not affect how much logs from other clusters are stored.

Right, yea it's definitely an option. But this feels like we might want to create a issue to implement support for this? Since this feels like a bigger task

**Comment by viktor-f** [comment]
> Right, yea it's definitely an option. But this feels like we might want to create a issue to implement support for this? Since this feels like a bigger task

I agree, an investigation issue to see if it's anything we would want.

**Review by viktor-f — APPROVED** [review]
